File: _Todo/377_Combination-Sum-IV/377_Combination-Sum-IV.cpp
/******************************************************************************
 * Title:
 *     377_Combination-Sum-IV
 *     377. 组合总和 Ⅳ
 * Address:
 *     https://leetcode-cn.com/problems/combination-sum-iv/
******************************************************************************/

#include <iostream>
#include <vector>
using namespace std;

// 【完全背包】

// 方法一：
class Solution {
public:
    int combinationSum4(vector<int>& nums, int target) {
        vector<unsigned long long int> dp( target + 1, 0 );
        dp[0] = 1;

        for ( int i = 0; i <= target; ++i ) {
            for ( int num : nums ) {
                if ( num <= i )
                    dp[i] += dp[i - num];
            }
        }

        return dp[target];
    }
};

int main()
{
    

    return 0;
}

File: _Todo/377_Combination-Sum-IV/377_Combination-Sum-IV.py
"""
 * Title:
 *     377_Combination-Sum-IV
 *     377. 组合总和 Ⅳ
 * Address:
 *     https://leetcode-cn.com/problems/combination-sum-iv/
"""

"""
    【完全背包】
"""

class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        dp = [0] * (target + 1)
        dp[0] = 1

        for i in range(target + 1):
            for num in nums:
                if num <= i:
                    dp[i] += dp[i - num]
        
        return dp[target]